gait3d/gaitanalysis_with_top_cam.py

```diff
@@ -0,0 +1,996 @@
+            closed_corner_opposite_side = f'MazeCornerClosed{opposite_side_id}'
+            angle_error = self._compute_angle_error(a = corners_and_likelihoods[f'MazeCornerClosed{opposite_side_id}']['coords'],
+                                              b = corners_and_likelihoods[f'MazeCornerClosed{side_id}']['coords'],
+                                              c = corners_and_likelihoods[open_corner_marker_id]['coords'])
+            distance_ratio_error = self._compute_distance_ratio_error(corners_and_likelihoods = corners_and_likelihoods,
+                                                                open_corner_marker_id = open_corner_marker_id,
+                                                                side_id = side_id)
+            if (angle_error <= tolerance) & (distance_ratio_error <= tolerance):
+                valid_positions = True
+            mean_error = (angle_error + distance_ratio_error) / 2
+            if mean_error < best_result['mean_error']:
+                best_result['valid'] = valid_positions
+                best_result['mean_error'] = mean_error
+                best_result['open_corner_id'] = open_corner_marker_id
+                best_result['side_id'] = side_id
+        return best_result
+    
+    
+    def _compute_angle_error(self, a: np.ndarray, b: np.ndarray, c: np.ndarray) -> float:
+        # b is point at the joint that connects the other two
+        ba = a - b
+        bc = c - b
+        cosine_angle = np.dot(ba, bc) / (np.linalg.norm(ba) * np.linalg.norm(bc))
+        angle = np.degrees(np.arccos(cosine_angle))
+        return self._compute_error_proportion(query_value = angle, target_value = 90)
+
+
+    def _compute_error_proportion(self, query_value: float, target_value: float) -> float:
+        return abs(query_value - target_value) / target_value
+
+
+    def _compute_distance_ratio_error(self, corners_and_likelihoods: Dict, open_corner_marker_id: str, side_id: str) -> float:
+        maze_width = self._get_distance_between_two_points(corners_and_likelihoods['MazeCornerClosedLeft']['coords'],
+                                                           corners_and_likelihoods['MazeCornerClosedRight']['coords'])
+        maze_length = self._get_distance_between_two_points(corners_and_likelihoods[f'MazeCornerClosed{side_id}']['coords'],
+                                                           corners_and_likelihoods[open_corner_marker_id]['coords'])
+        distance_ratio = maze_length/maze_width
+        return self._compute_error_proportion(query_value = distance_ratio, target_value = 50/4)
+
+
+    def _get_distance_between_two_points(self, coords_point_a: np.ndarray, coords_point_b: np.ndarray) -> float:
+        return ((coords_point_a[0] - coords_point_b[0])**2 + (coords_point_a[1] - coords_point_b[1])**2)**0.5
+
+
+    def _get_conversion_factor_px_to_cm(self, coords_point_a: np.ndarray, coords_point_b: np.ndarray, distance_in_cm: float) -> float:
```

Comment:
<img width="456" alt="image" src="https://user-images.githubusercontent.com/104254966/201874486-20433c01-7f4e-4165-9806-7a933801e51d.png">

Difference in pixel_length used as conversion factor based on detection of closed corners (c, n = 6) or open corners (o, n = 18)
Difference of means = 0.0033 px/cm
Difference of max and min conversion factor using closed corners is 0.021 px/cm

Conclusions: 
- conversion based on closed corners seems more stable between recordings and should be preferentially used to calculate conversion
- does conversion to cm differ too much between recordings?

Reply:
Very interesting! Would you suggest we switch to use the closed corners as default and only use an Open Corner if the relative error is very small? Right now the default is set to 25 %

I´ll keep the current version for now, but also raised an [issue](https://github.com/Defense-Circuits-Lab/Gait_Analysis/issues/41) for this topic, good catch! Maybe you could add the plots and information also there again.